File: src/ecs/world.ts
/**
 * The World: entity table, archetype graph, and component access.
 *
 * Layout, top to bottom:
 *
 *   Entity handle  ->  (archetype, row)          this file's `archetypeOf`/`rowOf`
 *   Archetype      ->  one column per component  archetype.ts
 *   Column         ->  one typed array per field component.ts
 *
 * Adding or removing a component MOVES the entity to a different archetype,
 * which is the expensive operation in this design (it copies every shared field
 * and swap-removes the old row). Components that toggle every few ticks should
 * therefore be a numeric field or a bool flag on a component the entity always
 * has — not their own tag. Components that describe what an entity fundamentally
 * *is* (a projectile, a pet, a centipede segment) are what tags are for.
 */

import {
    Archetype,
    archetypeKey,
    ComponentMask,
    createMask,
    maskContainsAll,
    maskIntersects,
    maskSet,
} from './archetype';
import {
    componentById,
    ComponentType,
    Columns,
    componentCount,
    Schema,
} from './component';
import {
    Entity,
    ENTITY_MAX_GENERATION,
    entityGeneration,
    entityIndex,
    entityToString,
    makeEntity,
    NULL_ENTITY,
} from './entity';

/** Initial number of entity slots; grows geometrically. */
const INITIAL_SLOTS = 1024;

/**
 * Values used to initialise a component when it is added.
 *
 * Deliberately `unknown` per field rather than a union of the storable
 * primitives: `obj` columns legitimately hold arbitrary payloads (loadouts,
 * Maps, config objects) that callers often have only as `unknown`, and a
 * narrower type here would force a cast at every prefab call site without
 * catching anything — the field's storage type already decides how the value is
 * used.
 */
export type ComponentInit<S extends Schema> = { [K in keyof S]?: unknown };

/**
 * One archetype's worth of rows handed to a system.
 *
 * `count` is the number of live rows; iterate `0 <= i < count`. `entities[i]`
 * is the handle for row `i`. Column arrays are fetched once per chunk and then
 * indexed directly — that hoisting is the whole point of the chunked API, so
 * always pull the arrays out ABOVE the row loop, never inside it.
 */
export interface Chunk {
    readonly archetype: Archetype;
    readonly count: number;
    readonly entities: Float64Array;
    /** The backing arrays for `type` in this chunk. Throws if not present. */
    cols<S extends Schema>(type: ComponentType<S>): Columns<S>;
}

class ChunkView implements Chunk {
    archetype!: Archetype;
    count = 0;
    entities!: Float64Array;

    cols<S extends Schema>(type: ComponentType<S>): Columns<S> {
        const col = this.archetype.columns[type.id];
        if (col === undefined) {
            throw new Error(
                `Chunk does not contain component "${type.name}". ` +
                `Add it to the query, or use an optional-component branch.`,
            );
        }
        return col.arrays as Columns<S>;
    }
}

/**
 * A cached set of archetypes matching a component filter.
 *
 * Queries are created once (ideally at module scope or system construction) and
 * re-iterated every tick. The matching archetype list is rebuilt only when the
 * world grows a new archetype, so steady-state iteration costs nothing beyond
 * the loop itself.
 */
export class Query {
    private readonly all: ComponentMask;
    private readonly none: ComponentMask;
    private readonly matched: Archetype[] = [];
    private seenArchetypes = -1;

    constructor(
        private readonly world: World,
        all: ReadonlyArray<ComponentType<any>>,
        none: ReadonlyArray<ComponentType<any>> = [],
    ) {
        const cap = Math.max(componentCount(), 1);
        this.all = createMask(cap);
        this.none = createMask(cap);
        for (const t of all) maskSet(this.all, t.id);
        for (const t of none) maskSet(this.none, t.id);
    }

    /** Refresh the archetype list if the world has grown new archetypes. */
    private refresh(): void {
        const archetypes = this.world.archetypes;
        if (this.seenArchetypes === archetypes.length) return;
        this.matched.length = 0;
        for (const a of archetypes) {
            if (maskContainsAll(a.mask, this.all) && !maskIntersects(a.mask, this.none)) {
                this.matched.push(a);
            }
        }
        this.seenArchetypes = archetypes.length;
    }

    /**
     * Run `fn` over every matching chunk.
     *
     * The chunk view is REUSED between calls to avoid per-tick allocation, so
     * never retain it past the callback.
     */
    chunks(fn: (chunk: Chunk) => void): void {
        this.refresh();
        const view = this.world.chunkView;
        for (const a of this.matched) {
            if (a.count === 0) continue;
            view.archetype = a;
            view.count = a.count;
            view.entities = a.entities;
            fn(view);
        }
    }

    /**
     * Collect matching entities into an array.
     *
     * Allocates, so it is for setup/teardown/debug paths and for the "destroy
     * everything matching" case — never a per-tick hot loop. Taking a snapshot
     * first is also what makes structural mutation safe while acting on results.
     */
    collect(out: Entity[] = []): Entity[] {
        out.length = 0;
        this.refresh();
        for (const a of this.matched) {
            for (let i = 0; i < a.count; i++) out.push(a.entities[i] as Entity);
        }
        return out;
    }

    /** Number of entities currently matching. */
    count(): number {
        this.refresh();
        let n = 0;
        for (const a of this.matched) n += a.count;
        return n;
    }
}

export class World {
    // --- entity table --------------------------------------------------------
    private generations = new Uint32Array(INITIAL_SLOTS);
    private alive = new Uint8Array(INITIAL_SLOTS);
    private archetypeOf = new Int32Array(INITIAL_SLOTS).fill(-1);
    private rowOf = new Int32Array(INITIAL_SLOTS).fill(-1);
    private freeSlots: number[] = [];
    /**
     * Starts at 1: slot 0 is permanently reserved and never handed out, because
     * index 0 with generation 0 packs to the handle `0` — which is NULL_ENTITY.
     * Without this the very first entity created would be indistinguishable from
     * "no entity", and every zero-filled `entity` column would appear to point
     * at it.
     */
    private slotCount = 1;
    private capacity = INITIAL_SLOTS;

    // --- archetypes ----------------------------------------------------------
    readonly archetypes: Archetype[] = [];
    private readonly archetypeByKey = new Map<string, Archetype>();

    /** Reused chunk view; see Query.chunks. */
    readonly chunkView = new ChunkView();

    /** Live entity count. */
    private liveCount = 0;

    /**
     * String id <-> entity, for the parts of the game that address entities by
     * socket id or mob id (the wire protocol, squads, pet ownership, targeting).
     * Kept here rather than in a component so `destroy` can drop both directions
     * in one place and never leak an id.
     */
    private readonly byExternalId = new Map<string, Entity>();
    private readonly externalIds: string[] = [];

    constructor() {
        // The empty archetype (no components) holds freshly created entities.
        this.getOrCreateArchetype([]);
    }

    // -------------------------------------------------------------------------
    // Entity lifecycle
    // -------------------------------------------------------------------------

    /** Create an entity with no components. */
    create(): Entity {
        let index: number;
        if (this.freeSlots.length > 0) {
            index = this.freeSlots.pop()!;
        } else {
            index = this.slotCount++;
            if (index >= this.capacity) this.growSlots(this.capacity * 2);
        }

        const entity = makeEntity(index, this.generations[index]);
        this.alive[index] = 1;
        this.liveCount++;

        const empty = this.archetypes[0];
        this.archetypeOf[index] = 0;
        this.rowOf[index] = empty.addRow(entity);
        return entity;
    }

    /** True when the handle refers to a live entity (right generation). */
    isAlive(e: Entity): boolean {
        if (e === NULL_ENTITY) return false;
        const index = entityIndex(e);
        return index < this.slotCount
            && this.alive[index] === 1
            && this.generations[index] === entityGeneration(e);
    }

    /**
     * Destroy an entity and free its slot.
     *
     * Bumps the slot's generation so every outstanding handle to it goes stale.
     * Safe to call on an already-dead handle (returns false).
     */
    destroy(e: Entity): boolean {
        if (!this.isAlive(e)) return false;
        const index = entityIndex(e);

        const archetype = this.archetypes[this.archetypeOf[index]];
        this.releaseRow(archetype, this.rowOf[index]);

        const externalId = this.externalIds[index];
        if (externalId !== undefined) {
            this.byExternalId.delete(externalId);
            this.externalIds[index] = undefined as unknown as string;
        }

        this.alive[index] = 0;
        this.archetypeOf[index] = -1;
        this.rowOf[index] = -1;
        // Wrap rather than overflow the Uint32Array; see entity.ts on why the
        // generation space is large enough that this is not a practical concern.
        this.generations[index] = (this.generations[index] + 1) % ENTITY_MAX_GENERATION;
        this.freeSlots.push(index);
        this.liveCount--;
        return true;
    }

    /** Number of live entities. */
    size(): number {
        return this.liveCount;
    }

    // -------------------------------------------------------------------------
    // Components
    // -------------------------------------------------------------------------

    /** True when `e` currently has `type`. */
    has(e: Entity, type: ComponentType<any>): boolean {
        if (!this.isAlive(e)) return false;
        return this.archetypes[this.archetypeOf[entityIndex(e)]].has(type.id);
    }

    /**
     * Add `type` to `e`, initialising the given fields (others start zeroed).
     * No-op apart from applying `values` when the entity already has it.
     */
    add<S extends Schema>(e: Entity, type: ComponentType<S>, values?: ComponentInit<S>): void {
        this.assertAlive(e, 'add');
        const index = entityIndex(e);
        const from = this.archetypes[this.archetypeOf[index]];

        if (!from.has(type.id)) {
            const ids = from.componentIds.slice();
            ids.push(type.id);
            const to = this.getOrCreateArchetypeByIds(ids);
            this.moveEntity(e, index, from, to);
            // Zero the destination slot: the row may be recycled storage.
            const col = to.columns[type.id]!;
            const row = this.rowOf[index];
            for (const field of type.fields) {
                const arr = col.arrays[field] as any;
                arr[row] = type.schema[field] === 'obj' || type.schema[field] === 'str' ? undefined : 0;
            }
        }

        if (values) this.write(e, type, values);
    }

    /** Remove `type` from `e`. No-op when absent. */
    remove(e: Entity, type: ComponentType<any>): void {
        this.assertAlive(e, 'remove');
        const index = entityIndex(e);
        const from = this.archetypes[this.archetypeOf[index]];
        if (!from.has(type.id)) return;

        const ids = from.componentIds.filter(id => id !== type.id);
        const to = this.getOrCreateArchetypeByIds(ids);
        this.moveEntity(e, index, from, to);
    }

    /** Read one field. Throws if the entity lacks the component. */
    get<S extends Schema, K extends Extract<keyof S, string>>(
        e: Entity,
        type: ComponentType<S>,
        field: K,
    ): any {
        this.assertAlive(e, 'get');
        const index = entityIndex(e);
        const archetype = this.archetypes[this.archetypeOf[index]];
        const col = archetype.columns[type.id];
        if (!col) throw new Error(`${entityToString(e)} has no component "${type.name}"`);
        return (col.arrays[field] as any)[this.rowOf[index]];
    }

    /** Write one field. Throws if the entity lacks the component. */
    set<S extends Schema, K extends Extract<keyof S, string>>(
        e: Entity,
        type: ComponentType<S>,
        field: K,
        value: any,
    ): void {
        this.assertAlive(e, 'set');
        const index = entityIndex(e);
        const archetype = this.archetypes[this.archetypeOf[index]];
        const col = archetype.columns[type.id];
        if (!col) throw new Error(`${entityToString(e)} has no component "${type.name}"`);
        (col.arrays[field] as any)[this.rowOf[index]] = value;
    }

    /** Write several fields at once. */
    write<S extends Schema>(e: Entity, type: ComponentType<S>, values: ComponentInit<S>): void {
        this.assertAlive(e, 'write');
        const index = entityIndex(e);
        const archetype = this.archetypes[this.archetypeOf[index]];
        const col = archetype.columns[type.id];
        if (!col) throw new Error(`${entityToString(e)} has no component "${type.name}"`);
        const row = this.rowOf[index];
        for (const field in values) {
            const v = values[field];
            if (v === undefined) continue;
            (col.arrays[field] as any)[row] = v;
        }
    }

    /**
     * Read every field of a component into a plain object.
     * Allocates — for debug dumps, persistence and wire encoding, not hot loops.
     */
    snapshot<S extends Schema>(e: Entity, type: ComponentType<S>): Record<string, any> {
        this.assertAlive(e, 'snapshot');
        const index = entityIndex(e);
        const col = this.archetypes[this.archetypeOf[index]].columns[type.id];
        if (!col) throw new Error(`${entityToString(e)} has no component "${type.name}"`);
        const row = this.rowOf[index];
        const out: Record<string, any> = {};
        for (const field of type.fields) out[field] = (col.arrays[field] as any)[row];
        return out;
    }

    /** Every component type currently on `e`. Diagnostics only. */
    componentsOf(e: Entity): ComponentType<any>[] {
        this.assertAlive(e, 'componentsOf');
        const archetype = this.archetypes[this.archetypeOf[entityIndex(e)]];
        const out: ComponentType<any>[] = [];
        for (const col of archetype.columns) if (col) out.push(col.type);
        return out;
    }

    // -------------------------------------------------------------------------
    // Queries
    // -------------------------------------------------------------------------

    /**
     * Build a query for entities having all of `all` and none of `none`.
     * Create these ONCE and reuse; constructing one per tick defeats the cache.
     */
    query(all: ReadonlyArray<ComponentType<any>>, none: ReadonlyArray<ComponentType<any>> = []): Query {
        return new Query(this, all, none);
    }

    // -------------------------------------------------------------------------
    // External string ids
    // -------------------------------------------------------------------------

    /** Associate a string id (socket id, mob id) with an entity. */
    bindExternalId(e: Entity, id: string): void {
        this.assertAlive(e, 'bindExternalId');
        const index = entityIndex(e);
        const previous = this.externalIds[index];
        if (previous !== undefined) this.byExternalId.delete(previous);
        this.externalIds[index] = id;
        this.byExternalId.set(id, e);
    }

    /** Look up an entity by its string id, if it is still alive. */
    lookup(id: string): Entity | undefined {
        const e = this.byExternalId.get(id);
        if (e === undefined) return undefined;
        if (!this.isAlive(e)) {
            this.byExternalId.delete(id);
            return undefined;
        }
        return e;
    }

    /** The string id bound to `e`, if any. */
    externalIdOf(e: Entity): string | undefined {
        if (!this.isAlive(e)) return undefined;
        return this.externalIds[entityIndex(e)];
    }

    // -------------------------------------------------------------------------
    // Internals
    // -------------------------------------------------------------------------

    private assertAlive(e: Entity, op: string): void {
        if (!this.isAlive(e)) {
            throw new Error(`${op}() called on dead or invalid entity ${entityToString(e)}`);
        }
    }

    private growSlots(capacity: number): void {
        const generations = new Uint32Array(capacity);
        generations.set(this.generations);
        this.generations = generations;

        const alive = new Uint8Array(capacity);
        alive.set(this.alive);
        this.alive = alive;

        const archetypeOf = new Int32Array(capacity).fill(-1);
        archetypeOf.set(this.archetypeOf);
        this.archetypeOf = archetypeOf;

        const rowOf = new Int32Array(capacity).fill(-1);
        rowOf.set(this.rowOf);
        this.rowOf = rowOf;

        this.externalIds.length = capacity;
        this.capacity = capacity;
    }

    private getOrCreateArchetype(types: ReadonlyArray<ComponentType<any>>): Archetype {
        return this.getOrCreateArchetypeByIds(types.map(t => t.id));
    }

    private getOrCreateArchetypeByIds(ids: ReadonlyArray<number>): Archetype {
        const key = archetypeKey(ids);
        const existing = this.archetypeByKey.get(key);
        if (existing) return existing;

        const types: ComponentType<any>[] = [];
        for (const id of ids) {
            const t = componentByIdOrThrow(id);
            types.push(t);
        }
        const archetype = new Archetype(types, Math.max(componentCount(), 1));
        archetype.index = this.archetypes.length;
        this.archetypeByKey.set(key, archetype);
        this.archetypes.push(archetype);
        return archetype;
    }

    /**
     * Move `e` from one archetype to another, carrying over every component the
     * two have in common. Components only in `from` are dropped; components only
     * in `to` are left for the caller to initialise.
     */
    private moveEntity(e: Entity, index: number, from: Archetype, to: Archetype): void {
        const fromRow = this.rowOf[index];
        const toRow = to.addRow(e);

        for (const id of to.componentIds) {
            const src = from.columns[id];
            if (!src) continue;
            const dst = to.columns[id]!;
            for (const field of src.type.fields) {
                (dst.arrays[field] as any)[toRow] = (src.arrays[field] as any)[fromRow];
            }
        }

        this.archetypeOf[index] = to.index;
        this.rowOf[index] = toRow;
        this.releaseRow(from, fromRow);
    }

    /**
     * Swap-remove `row` from `archetype` and repair the location of whichever
     * entity was moved into the hole.
     */
    private releaseRow(archetype: Archetype, row: number): void {
        const moved = archetype.removeRow(row);
        if (moved !== NULL_ENTITY) {
            this.rowOf[entityIndex(moved)] = row;
        }
    }
}

function componentByIdOrThrow(id: number): ComponentType<any> {
    const t = componentById(id);
    if (!t) throw new Error(`Unknown component id ${id} — was defineComponent() called at module scope?`);
    return t;
}

File: cpp/shared/core/world.cpp
#include "world.h"

#include <algorithm>
#include <cstdlib>
#include <stdexcept>
#include <string>

namespace flr {

// ---------------------------------------------------------------------------
// entity.h
// ---------------------------------------------------------------------------

std::string entityToString(Entity e) {
    if (e == NULL_ENTITY) return "Entity(null)";
    return "Entity(" + std::to_string(entityIndex(e)) + "v" +
           std::to_string(entityGeneration(e)) + ")";
}

// ---------------------------------------------------------------------------
// Component registry
// ---------------------------------------------------------------------------

namespace {

// Archetype columns hold a `const ComponentInfo*` for the lifetime of the
// world, so entries must never move. A fixed array (rather than a vector that
// reallocates as later component types register themselves) is what makes
// those pointers stable.
struct Registry {
    ComponentInfo entries[kMaxComponents];
    std::size_t count = 0;
};

Registry& registry() {
    static Registry r;
    return r;
}

} // namespace

namespace detail {
ComponentId registerComponent(ComponentInfo info) {
    Registry& r = registry();
    if (r.count >= kMaxComponents) {
        throw std::runtime_error("component registry full; raise flr::kMaxComponents");
    }
    info.id = static_cast<ComponentId>(r.count);
    r.entries[r.count++] = info;
    return info.id;
}
} // namespace detail

std::size_t componentCount() { return registry().count; }

const ComponentInfo* componentById(ComponentId id) {
    Registry& r = registry();
    return id < r.count ? &r.entries[id] : nullptr;
}

// ---------------------------------------------------------------------------
// Archetype
// ---------------------------------------------------------------------------

namespace {

void* alignedAllocate(std::size_t bytes, std::size_t align) {
    if (bytes == 0) return nullptr;
    // aligned_alloc requires a size that is a multiple of the alignment.
    const std::size_t alignment = std::max<std::size_t>(align, alignof(std::max_align_t));
    const std::size_t rounded = ((bytes + alignment - 1) / alignment) * alignment;
    void* p = std::aligned_alloc(alignment, rounded);
    if (!p) throw std::bad_alloc();
    return p;
}

} // namespace

Archetype::Archetype(ComponentMask mask, std::vector<ComponentId> ids)
    : mask_(mask), ids_(std::move(ids)) {
    std::fill(std::begin(columnOf_), std::end(columnOf_), static_cast<std::int16_t>(-1));
    columns_.reserve(ids_.size());
    for (std::size_t i = 0; i < ids_.size(); ++i) {
        Column c;
        c.info = componentById(ids_[i]);
        columns_.push_back(c);
        columnOf_[ids_[i]] = static_cast<std::int16_t>(i);
    }
}

Archetype::~Archetype() {
    for (Column& c : columns_) {
        if (!c.data) continue;
        if (!c.info->trivial) {
            for (std::size_t row = 0; row < count_; ++row) c.info->destroy(c.at(row));
        }
        std::free(c.data);
    }
}

void Archetype::reserve(std::size_t capacity) {
    if (capacity <= capacity_) return;
    // Geometric growth from a small floor. Most archetypes hold a handful of
    // entities (one per player, one per boss), so starting at 8 keeps the long
    // tail cheap while the crowded ones still double.
    std::size_t next = capacity_ == 0 ? 8 : capacity_;
    while (next < capacity) next *= 2;

    for (Column& c : columns_) {
        auto* fresh = static_cast<std::byte*>(alignedAllocate(next * c.info->size, c.info->align));
        if (c.data) {
            if (c.info->trivial) {
                std::memcpy(fresh, c.data, count_ * c.info->size);
            } else {
                for (std::size_t row = 0; row < count_; ++row) {
                    c.info->moveConstruct(fresh + row * c.info->size, c.at(row));
                    c.info->destroy(c.at(row));
                }
            }
            std::free(c.data);
        }
        c.data = fresh;
    }
    entities_.resize(next);
    capacity_ = next;
}

std::size_t Archetype::addRow(Entity e) {
    if (count_ == capacity_) reserve(count_ + 1);
    const std::size_t row = count_++;
    entities_[row] = e;
    // Always default-CONSTRUCT, never memset.
    //
    // Zeroing looks like a harmless fast path for a trivially copyable type,
    // and it silently throws away every in-class initialiser: a component
    // declaring `double speedScale = 1.0` would arrive as 0.0, and a system
    // multiplying by it would quietly do nothing at all. The compiler emits a
    // trivial constructor for these types anyway, so the "fast path" bought
    // nothing to begin with.
    for (Column& c : columns_) c.info->defaultConstruct(c.at(row));
    return row;
}

Entity Archetype::removeRow(std::size_t row) {
    const std::size_t last = count_ - 1;
    Entity moved = NULL_ENTITY;

    if (row != last) {
        for (Column& c : columns_) {
            if (c.info->trivial) {
                std::memcpy(c.at(row), c.at(last), c.info->size);
            } else {
                c.info->moveAssign(c.at(row), c.at(last));
            }
        }
        entities_[row] = entities_[last];
        moved = entities_[row];
    }

    if (!columns_.empty()) {
        for (Column& c : columns_) {
            if (!c.info->trivial) c.info->destroy(c.at(last));
        }
    }
    count_ = last;
    return moved;
}

// ---------------------------------------------------------------------------
// World
// ---------------------------------------------------------------------------

namespace {

std::string archetypeKey(const std::vector<ComponentId>& ids) {
    std::string key;
    key.reserve(ids.size() * 4);
    for (const ComponentId id : ids) {
        key += std::to_string(id);
        key += ',';
    }
    return key;
}

} // namespace

World::World() {
    // Slot 0 is reserved so that no live handle can equal NULL_ENTITY.
    slots_.push_back(Slot{});
    names_.emplace_back();
    // The empty archetype holds freshly created, component-less entities.
    getOrCreateArchetype({});
}

World::~World() = default;

Entity World::create() {
    std::uint32_t index;
    if (!freeSlots_.empty()) {
        index = freeSlots_.back();
        freeSlots_.pop_back();
    } else {
        index = static_cast<std::uint32_t>(slots_.size());
        slots_.push_back(Slot{});
        names_.emplace_back();
    }

    Slot& slot = slots_[index];
    const Entity e = makeEntity(index, slot.generation);
    slot.alive = true;
    slot.archetype = 0;
    slot.row = static_cast<std::uint32_t>(archetypes_[0]->addRow(e));

    ++liveCount_;
    ++structuralVersion_;
    return e;
}

bool World::destroy(Entity e) {
    if (!isAlive(e)) return false;
    const std::uint32_t index = entityIndex(e);
    Slot& slot = slots_[index];

    releaseRow(archetypes_[slot.archetype].get(), slot.row);

    if (!names_[index].empty()) {
        byName_.erase(names_[index]);
        names_[index].clear();
    }

    slot.alive = false;
    // Wrapping the generation is fine: a handle only aliases after 2^32
    // recycles of the same slot, which no session reaches.
    ++slot.generation;

    freeSlots_.push_back(index);
    --liveCount_;
    ++structuralVersion_;
    return true;
}

void World::clear() {
    for (std::uint32_t i = 1; i < slots_.size(); ++i) {
        if (slots_[i].alive) destroy(makeEntity(i, slots_[i].generation));
    }
    byName_.clear();
}

void World::assertAlive(Entity e, const char* op) const {
    if (!isAlive(e)) {
        throw std::runtime_error(std::string(op) + "() on dead or invalid " + entityToString(e));
    }
}

Archetype* World::getOrCreateArchetype(std::vector<ComponentId> ids) {
    std::sort(ids.begin(), ids.end());
    const std::string key = archetypeKey(ids);
    if (auto it = archetypeByKey_.find(key); it != archetypeByKey_.end()) {
        return archetypes_[it->second].get();
    }

    ComponentMask mask;
    for (const ComponentId id : ids) mask.set(id);

    auto archetype = std::make_unique<Archetype>(mask, std::move(ids));
    archetype->index_ = static_cast<std::uint32_t>(archetypes_.size());
    Archetype* raw = archetype.get();
    archetypeByKey_.emplace(key, archetype->index_);
    archetypes_.push_back(std::move(archetype));
    return raw;
}

void World::moveEntity(Entity e, Archetype* from, Archetype* to) {
    Slot& slot = slots_[entityIndex(e)];
    const std::size_t fromRow = slot.row;
    const std::size_t toRow = to->addRow(e);

    // Carry over every component the two archetypes share. Components only in
    // `from` are dropped; components only in `to` keep their fresh default.
    for (const ComponentId id : to->componentIds()) {
        Column* src = from->column(id);
        if (!src) continue;
        Column* dst = to->column(id);
        if (src->info->trivial) {
            std::memcpy(dst->at(toRow), src->at(fromRow), src->info->size);
        } else {
            src->info->moveAssign(dst->at(toRow), src->at(fromRow));
        }
    }

    slot.archetype = to->index();
    slot.row = static_cast<std::uint32_t>(toRow);
    releaseRow(from, fromRow);
}

void World::releaseRow(Archetype* archetype, std::size_t row) {
    const Entity moved = archetype->removeRow(row);
    if (moved != NULL_ENTITY) {
        slots_[entityIndex(moved)].row = static_cast<std::uint32_t>(row);
    }
}

// -- names ------------------------------------------------------------------

void World::bindName(Entity e, std::string name) {
    assertAlive(e, "bindName");
    const std::uint32_t index = entityIndex(e);
    if (!names_[index].empty()) byName_.erase(names_[index]);

    // Rebinding a name that another entity still holds must clear it there
    // too. Otherwise both entities believe they own it, and destroying the
    // stale one erases the binding out from under the live one.
    if (auto it = byName_.find(name); it != byName_.end()) {
        const std::uint32_t previous = entityIndex(it->second);
        if (previous < names_.size()) names_[previous].clear();
    }

    byName_[name] = e;
    names_[index] = std::move(name);
}

Entity World::lookup(const std::string& name) const {
    auto it = byName_.find(name);
    if (it == byName_.end()) return NULL_ENTITY;
    // A stale binding can only happen if destroy() was bypassed; treat the
    // handle as authoritative rather than trusting the map.
    return isAlive(it->second) ? it->second : NULL_ENTITY;
}

const std::string* World::nameOf(Entity e) const {
    if (!isAlive(e)) return nullptr;
    const std::string& name = names_[entityIndex(e)];
    return name.empty() ? nullptr : &name;
}

} // namespace flr
